srml/balances/src/lib.rs

```diff
@@ -0,0 +1,1208 @@
+		// and the account information relevant to this subsystem is deleted (i.e. the
+		// account is reaped).
+		let outcome = if balance < T::ExistentialDeposit::get() {
+			Self::set_free_balance(who, balance);
+			UpdateBalanceOutcome::AccountKilled
+		} else {
+			if !<FreeBalance<T, I>>::exists(who) {
+				Self::new_account(&who, balance);
+			}
+			Self::set_free_balance(who, balance);
+			UpdateBalanceOutcome::Updated
+		};
+		(imbalance, outcome)
+	}
+}
+
+impl<T: Trait<I>, I: Instance> ReservableCurrency<T::AccountId> for Module<T, I>
+where
+	T::Balance: MaybeSerializeDeserialize + Debug,
+{
+	fn can_reserve(who: &T::AccountId, value: Self::Balance) -> bool {
+		Self::free_balance(who)
+			.checked_sub(&value)
+			.map_or(false, |new_balance| {
+				Self::ensure_can_withdraw(who, value, WithdrawReason::Reserve.into(), new_balance).is_ok()
+			})
+	}
+
+	fn slash_reserved(who: &T::AccountId, value: Self::Balance) -> (Self::NegativeImbalance, Self::Balance) {
+		let b = Self::reserved_balance(who);
+		let slash = cmp::min(b, value);
+		// underflow should never happen, but it if does, there's nothing to be done here.
+		Self::set_reserved_balance(who, b - slash);
+		(NegativeImbalance::new(slash), value - slash)
+	}
+
+	fn reserved_balance(who: &T::AccountId) -> Self::Balance {
+		<ReservedBalance<T, I>>::get(who)
+	}
+
+	fn reserve(who: &T::AccountId, value: Self::Balance) -> result::Result<(), &'static str> {
+		let b = Self::free_balance(who);
+		if b < value {
+			return Err("not enough free funds");
+		}
+		let new_balance = b - value;
+		Self::ensure_can_withdraw(who, value, WithdrawReason::Reserve.into(), new_balance)?;
+		Self::set_reserved_balance(who, Self::reserved_balance(who) + value);
+		Self::set_free_balance(who, new_balance);
+		Ok(())
+	}
+
+	fn unreserve(who: &T::AccountId, value: Self::Balance) -> Self::Balance {
+		let b = Self::reserved_balance(who);
+		let actual = cmp::min(b, value);
+		Self::set_free_balance(who, Self::free_balance(who) + actual);
+		Self::set_reserved_balance(who, b - actual);
+		value - actual
+	}
+
+	fn repatriate_reserved(
+		slashed: &T::AccountId,
+		beneficiary: &T::AccountId,
+		value: Self::Balance,
+	) -> result::Result<Self::Balance, &'static str> {
+		if Self::total_balance(beneficiary).is_zero() {
+			return Err("beneficiary account must pre-exist");
+		}
+		let b = Self::reserved_balance(slashed);
+		let slash = cmp::min(b, value);
+		Self::set_free_balance(beneficiary, Self::free_balance(beneficiary) + slash);
+		Self::set_reserved_balance(slashed, b - slash);
+		Ok(value - slash)
+	}
+}
+
+impl<T: Trait> LockableCurrency<T::AccountId> for Module<T>
+where
+	T::Balance: MaybeSerializeDeserialize + Debug,
+{
+	type Moment = T::Moment;
+
+	// `amount` > `free_balance` is allowed
+	fn set_lock(
+		id: LockIdentifier,
+		who: &T::AccountId,
+		amount: T::Balance,
+		until: Self::Moment,
+		reasons: WithdrawReasons,
+	) {
+		let now = <timestamp::Module<T>>::now();
```

Comment:
I found a problem that if the until is BlockNumber, it should not compare with a timestamp, right?

Reply:
In this case they’re `Moment`.